src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs

```diff
@@ -576,23 +589,46 @@ public ITableEntriesSnapshot GetSnapshot(int versionNumber)
 
             void IDisposable.Dispose()
             {
-                this.Presenter.AssertIsForeground();
+            }
 
-                // VS is letting go of us.  i.e. because a new FAR call is happening, or because
-                // of some other event (like the solution being closed).  Remove us from the set
-                // of sources for the window so that the existing data is cleared out.
-                Debug.Assert(_findReferencesWindow.Manager.Sources.Count == 1);
-                Debug.Assert(_findReferencesWindow.Manager.Sources[0] == this);
+            #endregion
 
-                _findReferencesWindow.Manager.RemoveSource(this);
+            private void RemoveSink(ITableDataSink sink)
+            {
+                ImmutableInterlocked.Update(ref _tableDataSinks, (sinks, arg) => sinks.Remove(arg), sink);
 
-                CancelSearch();
+                if (_tableDataSinks.IsEmpty)
+                {
+                    // TODO: Why is this a safe assumption? Where is the contract that these subscriptions are only disposed on the main thread?
+                    //       Even if VS always does it, other folks can add their own sinks, and if the requirement isn't documented,
+                    //       others may dispose on another thread.
+                    Presenter.AssertIsForeground();
+
+                    // VS is letting go of us.  i.e. because a new FAR call is happening, or because
+                    // of some other event (like the solution being closed).  Remove us from the set
+                    // of sources for the window so that the existing data is cleared out.
+                    _findReferencesWindow.Manager.RemoveSource(this);
 
-                // Remove ourselves from the list of contexts that are currently active.
-                Presenter._currentContexts.Remove(this);
+                    CancelSearch();
+
+                    // Remove ourselves from the list of contexts that are currently active.
+                    Presenter._currentContexts.Remove(this);
+                }
             }
 
-            #endregion
+            private class DisposableSubscription : IDisposable
+            {
+                private readonly AbstractTableDataSourceFindUsagesContext _owner;
+                private readonly ITableDataSink _sink;
+
+                internal DisposableSubscription(AbstractTableDataSourceFindUsagesContext owner, ITableDataSink sink)
+                {
+                    _owner = owner ?? throw new ArgumentNullException(nameof(owner));
+                    _sink = sink ?? throw new ArgumentNullException(nameof(sink));
+                }
+
+                public void Dispose() => _owner.RemoveSink(_sink);
```

Comment:
please place `=>` on the next line, indented.

Reply:
Yup. Another style I missed. Thanks.